src/legacy/ui/public/vis/vis_types/vislib_vis_type.js

```diff
@@ -92,6 +92,10 @@ export function VislibVisTypeProvider(Private, $rootScope, $timeout, $compile) {
         if (this.vis.params.addLegend) {
           this.$scope.refreshLegend++;
           this.$scope.$digest();
+
+          // re-render after the legend is initialized to correctly take
+          // the legend height into account
+          this.vis.vislibVis.render(esResponse, this.vis.getUiState());
```

Comment:
why was this working for the left/right legend ? we have the same scenario, where chart needs to take new width into account ?

Reply:
I think the left/right legend has a fixed width and is taken correctly into account even if there are no entries in the legend in the first render pass

Reply:
ok ... so before the logic was 
- render the legend (so chart knows the size)
- render the chart 
- render the legend again with correct data (which chart prepared)

now:
- render legend
- render chart
- render legend with correct data
- render chart with correct size

.... this feels a bit too much, but i agree its a working easy fix

couldn't we just:
- render chart
- render legend with correct data
- render chart

(this way we get rid of one legend render)
or even better, extract the logic in chart used to produce correct legend info, or separate the chart in two parts .. prepare data and render

then:
- chart prepare data
- render legend (with correct data)
- chart render

Reply:
> * chart prepare data
> * render legend (with correct data)
> * chart render
I tried to do this at first, but a lot of tests failed because of it. I will restore the state so we can take a look on it.